server/src/main/java/org/opensearch/search/startree/filter/provider/StarTreeFilterProvider.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.startree.filter.provider;
+
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.Query;
+import org.opensearch.common.annotation.ExperimentalApi;
+import org.opensearch.index.compositeindex.datacube.Dimension;
+import org.opensearch.index.mapper.CompositeDataCubeFieldType;
+import org.opensearch.index.mapper.MappedFieldType;
+import org.opensearch.index.query.MatchAllQueryBuilder;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.RangeQueryBuilder;
+import org.opensearch.index.query.TermQueryBuilder;
+import org.opensearch.index.query.TermsQueryBuilder;
+import org.opensearch.search.internal.SearchContext;
+import org.opensearch.search.startree.StarTreeQueryHelper;
+import org.opensearch.search.startree.filter.StarTreeFilter;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Converts a {@link QueryBuilder} into a {@link StarTreeFilter} by generating the appropriate @{@link org.opensearch.search.startree.filter.DimensionFilter}
+ * for the fields provided in the user query.
+ */
+@ExperimentalApi
+public interface StarTreeFilterProvider {
+
+    /**
+     * Returns the {@link StarTreeFilter} generated from the {@link QueryBuilder}
+     * @param context:
+     * @param rawFilter:
+     * @param compositeFieldType:
+     * @return : {@link StarTreeFilter} if the query shape is supported, else null.
+     * @throws IOException :
+     */
+    StarTreeFilter getFilter(SearchContext context, QueryBuilder rawFilter, CompositeDataCubeFieldType compositeFieldType)
+        throws IOException;
+
+    StarTreeFilterProvider MATCH_ALL_PROVIDER = (context, rawFilter, compositeFieldType) -> new StarTreeFilter(Collections.emptyMap());
+
+    /**
+     * Singleton instances for most {@link StarTreeFilterProvider}
+     */
+    class SingletonFactory {
+
+        private static final Map<Class<? extends QueryBuilder>, StarTreeFilterProvider> QUERY_BUILDERS_TO_STF_PROVIDER = Map.of(
+            MatchAllQueryBuilder.class,
+            MATCH_ALL_PROVIDER,
+            TermQueryBuilder.class,
+            new TermStarTreeFilterProvider(),
+            TermsQueryBuilder.class,
+            new TermsStarTreeFilterProvider(),
+            RangeQueryBuilder.class,
+            new RangeStarTreeFilterProvider()
+        );
```

Comment:
Since every implementation of `QueryBuilder` should have a uniquely-identifying `String` name (that gets used for both its `Writeable` and `XContent` format), could we just use that as a key instead?

In particular, to make this pluggable, we can use `QuerySpec#getName().getPreferredName()` as a key. Then we'd just need to add an optional `setStarTreeFilterProvider` method to `QuerySpec`. If a plugin exposes that, we could register the filter provider on startup. It's consistent with other pluggable logic.

Reply:
Changed this to use `QueryBuilder.name`